lib/flow/Advection.cpp

```diff
@@ -43,102 +43,146 @@ Advection::CheckReady() const
 }
 
 
-int
-Advection::AdvectOneStep( Field* velocity, float deltaT, ADVECTION_METHOD method )
+int Advection::AdvectSteps( Field* velocity, float deltaT, size_t maxSteps, ADVECTION_METHOD method )
 {
     int ready = CheckReady();
     if( ready != 0 )
         return ready;
-
     bool happened = false;
-    size_t streamIdx = 0;
-    for( auto& s : _streams )       // Process one stream at a time
+
+    // Observation: user parameters are not gonna change while this function executes.
+    // Action: lock these parameters.
+    if( velocity->LockParams() != 0 )
+        return PARAMS_ERROR;
+
+    // The particle advection process can be parallelized per particle
+    // Each stream represents a trajectory for a single particle
+    // #pragma omp parallel for
+    for(size_t streamIdx = 0; streamIdx < _streams.size(); streamIdx++)
     {
-        // Check if the particle is inside of the volume.
-        // Also wrap it along periodic dimensions if enabled.
-        if( !velocity->InsideVolumeVelocity( s.back().time, s.back().location ) )
+        auto& s = _streams[streamIdx];
+        size_t numberOfSteps = s.size() - _separatorCount[streamIdx];
+        while(numberOfSteps < maxSteps )
         {
-            auto& p0 = s.back();
-            // Attempt to apply periodicity
-            bool locChanged = false;
-            auto loc = p0.location;
-            for( int i = 0; i < 3; i++ )    // correct coordinates in each periodic dimension
-            {
-                if( _isPeriodic[i] )
+            auto& past0 = s.back();
+            if( past0.IsSpecial() ) // If the last particle is marked "special,"
+                break;              // terminate stream immediately.
+
+            float dt = deltaT;
+            if( s.size() > 2 )  // If there are at least 3 particles in the stream and
+            {                   // neither is a separator, we also adjust *dt*
+                const auto& past1 = s[ s.size()-2 ];
+                const auto& past2 = s[ s.size()-3 ];
+                if( (!past1.IsSpecial()) && (!past2.IsSpecial()) )
                 {
-                    loc[i] = _applyPeriodic( loc[i], _periodicBounds[i][0], _periodicBounds[i][1] );
-                    locChanged = true;
+                    float mindt = deltaT / 20.0f,   maxdt = deltaT * 20.0f;
```

Comment:
Please comment on what this does and the use of the literal 20.0. Of concern is whether assumptions are being made about the velocity field range

Reply:
more explanation is added.